core/src/main/java/net/automatalib/ts/modal/MembershipMC.java

```diff
@@ -0,0 +1,170 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package net.automatalib.ts.modal;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+import com.google.common.collect.UnmodifiableIterator;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.Set;
+
+import net.automatalib.automata.AutomatonCreator;
+import net.automatalib.automata.graphs.TransitionEdge;
+import net.automatalib.automata.graphs.UniversalAutomatonGraphView;
+import net.automatalib.automata.visualization.MMCVisualizationHelper;
+import net.automatalib.graphs.UniversalGraph;
+import net.automatalib.ts.modal.transitions.ModalContractEdgeProperty.EdgeColor;
+import net.automatalib.ts.modal.transitions.ModalContractMembershipEdgePropertyImpl;
+import net.automatalib.ts.modal.transitions.ModalEdgeProperty.ModalType;
+import net.automatalib.visualization.VisualizationHelper;
+import net.automatalib.words.Alphabet;
+import net.automatalib.words.GrowingAlphabet;
+import net.automatalib.words.impl.Alphabets;
+import net.automatalib.words.impl.ArrayAlphabet;
+import net.automatalib.words.impl.GrowingMapAlphabet;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+public class MembershipMC<I> extends AbstractCompactMTS<I, ModalContractMembershipEdgePropertyImpl>
+        implements MutableModalContract<Integer, I, MTSTransition<I, ModalContractMembershipEdgePropertyImpl>, ModalContractMembershipEdgePropertyImpl> {
+
+    protected LinkedHashSet<I> communicationAlphabet;
+
+    public MembershipMC(Alphabet<I> alphabet, Alphabet<I> gamma) {
+        super(alphabet);
+        this.communicationAlphabet = Sets.newLinkedHashSet(gamma);
+
+        assert new HashSet<>(alphabet).containsAll(gamma) : "Communication alphabet needs to be a subset of alphabet";
+    }
+
+    public MembershipMC(Alphabet<I> alphabet) {
+        super(alphabet);
+        this.communicationAlphabet = Sets.newLinkedHashSetWithExpectedSize(alphabet.size() / 2);
+    }
+
+    public boolean addCommunicationSymbol(I symbol) {
+        boolean wasNotPresent = communicationAlphabet.add(symbol);
+
+        assert getInputAlphabet().contains(symbol) : "Communication alphabet needs to be a subset of alphabet";
+
+        return wasNotPresent;
+    }
+
+    public void removeCommunicationSymbol(I symbol) {
+        communicationAlphabet.remove(symbol);
+    }
```

Comment:
Is this functionality really needed? Currently, I don't see this function used. Normally, the (mutable) automaton types only support incremental operations and removing things (e.g. states) is realized via additional interfaces (e.g. `ShrinkableAutomaton`).

Reply:
Yes, it is. Inside a heavily-iterated loop we use a pruning method which removes unused labels. This is also relevant for following steps, so reordering would change the semantics. Copying the whole modal contract inside the loop is to expensive, as the graph uses to much memory. Therefore we need an implementation with dynamic mutators (adding/shrinking).

Extracting the functionality into an interface is a good idea.

Reply:
Does this only concern alphabet symbols of the communication alphabet or the removal of states as well? (regarding a possible *shrinkable* interface)

Is it important to store this alphabet on the (M)MC instance or would it be possible/more universal to pass the communication-alphabet-to-consider as a method parameter (similar to how you can e.g. compute a shortest path using only the given set of input symbols)?

Reply:
Storing the alphabet separately is an option, but it would (probably) require changing many method signatures.

However, the "shrinked" alphabet is the communication alphabet. So when storing it separately it would always replace the intern communication alphabet, i.e. there is no semantic interpretation of the superset.

Other (shrinkable) methods may be used to further improve performance, but are currently not required.

Reply:
The main reason why I was worried is because with regular (input) alphabets, the symbol indexes are important for computing successors (e.g. each state has an array of length | \Sigma | which stores the transitions). So by removing input symbol this creates non-contiguous memory which we should prevent. However, so far it seems that this is only an additional property of an (M)MC and not related to any internal computations. For now, I'll leave it as-is, but we may discuss further possible cleanups in the future.